File: java/src/main/java/lambda/Load.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import saaf.Inspector;
import saaf.Response;

import java.io.*;
import java.nio.file.Paths;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.Statement;
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;

public class Load implements RequestHandler<Map<String, Object>, Map<String, Object>> {

    private static final AmazonS3 s3Client = AmazonS3ClientBuilder.defaultClient();
    private static final Logger logger = Logger.getLogger(Load.class.getName());
    private static final int BATCH_SIZE = 1000;
    private static final int BUFFER_SIZE = 8192;

    @Override
    public Map<String, Object> handleRequest(Map<String, Object> event, Context context) {
        Inspector inspector = new Inspector();
        inspector.inspectAll();

        logger.info("Event received: " + event);

        try {
            // Extract parameters
            String bucketName = (String) event.get("bucket_name");
            String csvFileKey = (String) event.get("csv_file_key");
            String dbFileName = (String) event.getOrDefault("db_file_name", "data.db");

            if (bucketName == null || csvFileKey == null) {
                throw new IllegalArgumentException("Both 'bucket_name' and 'csv_file_key' are required.");
            }

            inspector.addAttribute("bucket_name", bucketName);
            inspector.addAttribute("csv_file_key", csvFileKey);
            inspector.addAttribute("db_file_name", dbFileName);

            String localCsvPath = "/tmp/" + Paths.get(csvFileKey).getFileName();
            String localDbPath = "/tmp/" + dbFileName;

            // Download CSV using buffered streaming
            downloadCsvFromS3(bucketName, csvFileKey, localCsvPath);

            // Process CSV and create database
            processDataAndCreateDb(localCsvPath, localDbPath);

            // Upload database to S3
            String s3Key = "databases/" + dbFileName;
            uploadDatabaseToS3(localDbPath, bucketName, s3Key);

            // Prepare response
            Response response = new Response();
            response.setValue("Database created and uploaded successfully");
            inspector.addAttribute("s3_upload_key", s3Key);
            inspector.consumeResponse(response);

        } catch (Exception e) {
            logger.log(Level.SEVERE, "An error occurred", e);
            Response response = new Response();
            inspector.consumeResponse(response);
        } finally {
            // Cleanup temporary files
            //cleanupTempFiles(localCsvPath, localDbPath);
        }

        inspector.inspectAllDeltas();
        return inspector.finish();
    }

    private void downloadCsvFromS3(String bucketName, String csvFileKey, String localPath) throws IOException {
        logger.info("Downloading CSV file from S3: " + bucketName + "/" + csvFileKey);

        try (S3Object s3Object = s3Client.getObject(new GetObjectRequest(bucketName, csvFileKey));
             BufferedInputStream bis = new BufferedInputStream(s3Object.getObjectContent(), BUFFER_SIZE);
             BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(localPath), BUFFER_SIZE)) {

            byte[] buffer = new byte[BUFFER_SIZE];
            int bytesRead;
            while ((bytesRead = bis.read(buffer)) != -1) {
                bos.write(buffer, 0, bytesRead);
            }
        }
        logger.info("CSV file downloaded successfully");
    }

    private void processDataAndCreateDb(String csvPath, String dbPath) throws Exception {
        Class.forName("org.sqlite.JDBC");

        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath)) {
            conn.setAutoCommit(false);  // Enable transaction mode

            // Create table with optimized indices
            createTableWithIndices(conn);

            // Process CSV in batches
            try (BufferedReader br = new BufferedReader(new FileReader(csvPath), BUFFER_SIZE)) {
                br.readLine(); // Skip header
                processDataInBatches(br, conn);
            }
        }
    }

    private void createTableWithIndices(Connection conn) throws Exception {
        try (Statement stmt = conn.createStatement()) {
            // Create table with appropriate data types and constraints
            stmt.execute("""
                CREATE TABLE orders (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    Region TEXT NOT NULL,
                    Country TEXT NOT NULL,
                    ItemType TEXT NOT NULL,
                    SalesChannel TEXT NOT NULL,
                    OrderPriority TEXT NOT NULL,
                    OrderDate TEXT NOT NULL,
                    OrderID INTEGER NOT NULL,
                    ShipDate TEXT NOT NULL,
                    UnitsSold INTEGER NOT NULL,
                    UnitPrice REAL NOT NULL,
                    UnitCost REAL NOT NULL,
                    TotalRevenue REAL NOT NULL,
                    TotalCost REAL NOT NULL,
                    TotalProfit REAL NOT NULL,
                    OrderProcessingTime INTEGER NOT NULL,
                    GrossMargin REAL NOT NULL
                );
                
                CREATE INDEX idx_orderid ON orders(OrderID);
                CREATE INDEX idx_region_country ON orders(Region, Country);
                CREATE INDEX idx_orderdate ON orders(OrderDate);
            """);
            conn.commit();
        }
    }

    private void processDataInBatches(BufferedReader br, Connection conn) throws Exception {
        String insertQuery = """
            INSERT INTO orders (
                Region, Country, ItemType, SalesChannel, OrderPriority,
                OrderDate, OrderID, ShipDate, UnitsSold, UnitPrice, UnitCost,
                TotalRevenue, TotalCost, TotalProfit, OrderProcessingTime, GrossMargin
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
        """;

        try (PreparedStatement pstmt = conn.prepareStatement(insertQuery)) {
            String line;
            int count = 0;

            while ((line = br.readLine()) != null) {
                String[] values = line.split(",");
                bindParameters(pstmt, values);
                pstmt.addBatch();
                count++;

                if (count % BATCH_SIZE == 0) {
                    pstmt.executeBatch();
                    conn.commit();
                    pstmt.clearBatch();
                }
            }

            // Execute remaining batch
            if (count % BATCH_SIZE != 0) {
                pstmt.executeBatch();
                conn.commit();
            }
        }
    }

    private void bindParameters(PreparedStatement pstmt, String[] values) throws Exception {
        pstmt.setString(1, values[0]);  // Region
        pstmt.setString(2, values[1]);  // Country
        pstmt.setString(3, values[2]);  // ItemType
        pstmt.setString(4, values[3]);  // SalesChannel
        pstmt.setString(5, values[4]);  // OrderPriority
        pstmt.setString(6, values[5]);  // OrderDate
        pstmt.setInt(7, Integer.parseInt(values[6]));  // OrderID
        pstmt.setString(8, values[7]);  // ShipDate
        pstmt.setInt(9, Integer.parseInt(values[8]));  // UnitsSold
        pstmt.setDouble(10, Double.parseDouble(values[9]));  // UnitPrice
        pstmt.setDouble(11, Double.parseDouble(values[10])); // UnitCost
        pstmt.setDouble(12, Double.parseDouble(values[11])); // TotalRevenue
        pstmt.setDouble(13, Double.parseDouble(values[12])); // TotalCost
        pstmt.setDouble(14, Double.parseDouble(values[13])); // TotalProfit
        pstmt.setInt(15, Integer.parseInt(values[14]));      // OrderProcessingTime
        pstmt.setDouble(16, Double.parseDouble(values[15])); // GrossMargin
    }

    private void uploadDatabaseToS3(String dbPath, String bucketName, String s3Key) throws IOException {
        logger.info("Uploading SQLite database to S3: " + bucketName + "/" + s3Key);

        try (BufferedInputStream dbFileStream = new BufferedInputStream(new FileInputStream(dbPath), BUFFER_SIZE)) {
            s3Client.putObject(new PutObjectRequest(bucketName, s3Key, dbFileStream, null));
        }
        logger.info("Database uploaded successfully");
    }

    private void cleanupTempFiles(String... filePaths) {
        for (String path : filePaths) {
            try {
                new File(path).delete();
            } catch (Exception e) {
                logger.warning("Failed to delete temporary file: " + path);
            }
        }
    }
}

File: python/src/Load.py
import boto3
import os
import sqlite3
import csv
import json
import logging
from Inspector import Inspector  # Ensure the Inspector class is available in your Lambda deployment package

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# Initialize S3 client
s3_client = boto3.client('s3')

def lambda_handler(event, context):
    inspector = Inspector()
    inspector.inspectAll()  # Start collecting runtime and system metrics

    try:
        logger.info("Lambda function started.")
        logger.info(f"Received event: {json.dumps(event, indent=2)}")

        # Extract parameters from the event
        bucket_name = event.get("bucket_name")
        csv_file_key = event.get("key")
        db_file_name = event.get("db_file_name", "data.db")  # Default file name: data.db

        if not bucket_name or not csv_file_key:
            raise ValueError("Both 'bucket_name' and 'csv_file_key' are required.")

        # Paths for local files
        local_csv_path = f"/tmp/{os.path.basename(csv_file_key)}"
        local_db_path = f"/tmp/{db_file_name}"

        # Download the CSV file from S3
        logger.info(f"Downloading CSV file from S3: {bucket_name}/{csv_file_key}")
        s3_client.download_file(bucket_name, csv_file_key, local_csv_path)
        logger.info(f"CSV file downloaded successfully to {local_csv_path}")

        # Check if the SQLite database exists; if not, create it
        conn = sqlite3.connect(local_db_path)
        cursor = conn.cursor()

        # Create a table if it doesn't exist
        create_table_query = '''
            CREATE TABLE IF NOT EXISTS orders (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                Region TEXT,
                Country TEXT,
                ItemType TEXT,
                SalesChannel TEXT,
                OrderPriority TEXT,
                OrderDate TEXT,
                OrderID INTEGER,
                ShipDate TEXT,
                UnitsSold INTEGER,
                UnitPrice REAL,
                UnitCost REAL,
                TotalRevenue REAL,
                TotalCost REAL,
                TotalProfit REAL,
                OrderProcessingTime INTEGER,
                GrossMargin REAL
            );
        '''
        cursor.execute(create_table_query)
        conn.commit()

        # Read and insert CSV data into the SQLite database
        logger.info(f"Reading CSV file and inserting data into SQLite database: {local_db_path}")
        with open(local_csv_path, newline='', mode='r') as csvfile:
            reader = csv.DictReader(csvfile)
            insert_query = '''
                INSERT INTO orders (
                    Region, Country, ItemType, SalesChannel, OrderPriority,
                    OrderDate, OrderID, ShipDate, UnitsSold, UnitPrice,
                    UnitCost, TotalRevenue, TotalCost, TotalProfit,
                    OrderProcessingTime, GrossMargin
                )
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            '''
            data_to_insert = [
                (
                    row['Region'], row['Country'], row['Item Type'], row['Sales Channel'], row['Order Priority'],
                    row['Order Date'], int(row['Order ID']), row['Ship Date'], int(row['Units Sold']), float(row['Unit Price']),
                    float(row['Unit Cost']), float(row['Total Revenue']), float(row['Total Cost']), float(row['Total Profit']),
                    int(row['Order Processing Time']), float(row['Gross Margin'])
                ) for row in reader
            ]
            cursor.executemany(insert_query, data_to_insert)
        conn.commit()
        conn.close()
        logger.info(f"Data inserted successfully into SQLite database.")

        # Upload the SQLite database to S3
        s3_key = f"databases/{db_file_name}"  # S3 key for the database file
        logger.info(f"Uploading SQLite database to S3: {bucket_name}/{s3_key}")

        with open(local_db_path, "rb") as db_file:
            s3_client.put_object(Body=db_file, Bucket=bucket_name, Key=s3_key)

        logger.info("Database uploaded successfully.")

        inspector.inspectAllDeltas()  # Collect deltas for runtime metrics
        runtime_metrics = inspector.finish()  # Finalize the runtime metrics collection

        # Merge runtime_metrics into the top-level response
        response = {
            "statusCode": 200,
            "message": "Database created and uploaded successfully",
            "bucket_name": bucket_name,
            "s3_key": s3_key
        }
        response.update(runtime_metrics)  # Add runtime metrics to the top level

        logger.info(f"Response: {response}")
        return response

    except Exception as e:
        inspector.inspectAllDeltas()  # Collect deltas even on failure
        runtime_metrics = inspector.finish()  # Finalize the runtime metrics collection
        logger.error(f"An error occurred: {e}", exc_info=True)
        error_response = {
            "statusCode": 500,
            "error": str(e)
        }
        error_response.update(runtime_metrics)  # Add runtime metrics to the error response
        return error_response

